File: da1/da1/move.py
#!/usr/bin/env python3

import rclpy 
from rclpy.node import Node
from da1.zlac8015d import Controller
from std_msgs.msg import Float64MultiArray

# import numpy as np

class Move (Node):
    def __init__ (self):
        super().__init__("move")

        self.motors = Controller (port='/dev/ttyUSB1')
        self.motors.disable_motor()
        self.motors.set_accel_time(500, 500)
        self.motors.set_decel_time(500, 500)

        self.motors.set_mode(3)
        self.motors.enable_motor()

        self.move_sub = self.create_subscription (Float64MultiArray, "/PT/move", self.move_sub_callback, 10)
        self.get_logger().info("move soon")

    def move_sub_callback (self, msg: Float64MultiArray):

        self.motors.set_rpm (int(msg.data[0]), int(msg.data[1]))
        
        # print(f"rpmL-rpmL ouput: {l_rpm: .4f}, {-r_rpm: .4f}")

def main (args=None):
    rclpy.init(args=args)
    node = Move()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.motors.disable_motor()
    node.destroy_node()
    rclpy.shutdown()
if __name__ == '__main__':
    main()

File: stage0_ps4_control/src/move.cpp
#include <chrono>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float64_multi_array.hpp"
#include "zlac8015d_hw/zlac8015d.hpp"

using namespace std::chrono_literals;

class Move : public rclcpp::Node
{
public:
  Move() : Node("move"), motors_("/dev/ttyUSB1")
  {
    motors_.disableMotor();
    motors_.setAccelTime(500, 500);
    motors_.setDecelTime(500, 500);
    motors_.setMode(3);
    motors_.setRpm(0, 0);
    motors_.enableMotor();

    move_sub_ = create_subscription<std_msgs::msg::Float64MultiArray>(
      "/PT/move", 1, std::bind(&Move::moveCallback, this, std::placeholders::_1));

    rpm_timer_ = create_wall_timer(200ms, std::bind(&Move::readRpm, this));
    watchdog_timer_ = create_wall_timer(100ms, std::bind(&Move::checkWatchdog, this));

    last_move_time_ = now();

    RCLCPP_INFO(get_logger(), "move started");
  }

  ~Move()
  {
    try { motors_.setRpm(0, 0); } catch (...) {}
    try { motors_.disableMotor(); } catch (...) {}
  }

private:
  void moveCallback(const std_msgs::msg::Float64MultiArray::SharedPtr msg)
  {
    if (msg->data.size() < 2) {
      RCLCPP_WARN(get_logger(), "Invalid /PT/move message");
      return;
    }

    int left_rpm = static_cast<int>(msg->data[0]);
    int right_rpm = static_cast<int>(msg->data[1]);

    motors_.setRpm(left_rpm, right_rpm);

    last_move_time_ = now();
    move_received_ = true;
    watchdog_stopped_ = false;
  }

  void checkWatchdog()
  {
    if (!move_received_) return;

    double elapsed = (now() - last_move_time_).seconds();

    if (elapsed > command_timeout_ && !watchdog_stopped_) {
      motors_.setRpm(0, 0);
      watchdog_stopped_ = true;
      RCLCPP_WARN(get_logger(), "Move command timeout. Motors stopped.");
    }
  }

  void readRpm()
  {
    try {
      auto [left_rpm, right_rpm] = motors_.getRpm();

      RCLCPP_INFO(get_logger(), "RPM Left: %.4f | RPM Right: %.4f",
                  static_cast<double>(left_rpm), static_cast<double>(right_rpm));
    }
    catch (const std::exception & e) {
      RCLCPP_WARN(get_logger(), "Get RPM failed: %s", e.what());
    }
  }

  static constexpr double command_timeout_ = 0.5;

  bool move_received_ = false;
  bool watchdog_stopped_ = false;

  rclcpp::Time last_move_time_;

  zlac8015d_hw::Controller motors_;
  rclcpp::Subscription<std_msgs::msg::Float64MultiArray>::SharedPtr move_sub_;
  rclcpp::TimerBase::SharedPtr rpm_timer_, watchdog_timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Move>());
  rclcpp::shutdown();
  return 0;
}